File: packages/gbnf-experiment/data/20260909T151143Z_170b737d/ported_implementation/src/utils/errors/grammar_parse_error.ts
import { build_error_position } from "./build_error_position.ts";

export const GRAMMAR_PARSER_ERROR_HEADER_MESSAGE = (reason: string): string =>
  `Failed to parse grammar: ${reason}`;

export class GrammarParseError extends Error {
  grammar: string;
  pos: number;
  reason: string;

  constructor(grammar: string, pos: number, reason: string) {
    super(
      [
        GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
        "",
        ...build_error_position(grammar, pos),
      ].join("\n"),
    );
    this.name = "GrammarParseError";
    this.grammar = grammar;
    this.pos = pos;
    this.reason = reason;
  }

  toString(): string {
    return [
      GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(this.reason),
      "",
      ...build_error_position(this.grammar, this.pos),
    ].join("\n");
  }
}

File: packages/gbnf-experiment/data/20260909T040730Z_0947f229/ported_implementation/gbnf/utils/errors/grammar_parse_error.py
"""Port of ``src/utils/errors/grammar-parse-error.ts``."""

from __future__ import annotations

from .build_error_position import build_error_position


def GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason: str) -> str:
    return f'Failed to parse grammar: {reason}'


class GrammarParseError(Exception):
    def __init__(self, grammar: str, pos: int, reason: str) -> None:
        message = '\n'.join([
            GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
            '',
            *build_error_position(grammar, pos),
        ])
        super().__init__(message)
        self.name = 'GrammarParseError'
        self.message = message
        self.grammar = grammar
        self.reason = reason
        self.pos = pos
